File: PandorasBox/Ahorn/entities/coloredwater.jl
module PandorasBoxColoredWater

using ..Ahorn, Maple

@mapdef Entity "pandorasBox/coloredWater" ColoredWater(x::Integer, y::Integer, width::Integer=Maple.defaultBlockWidth, height::Integer=Maple.defaultBlockHeight, hasTop::Bool=true, hasBottom::Bool=false, hasLeft::Bool=false, hasRight::Bool=false, color::String="LightSkyBlue")

function getColor(color)
    if haskey(Ahorn.XNAColors.colors, color)
        return Ahorn.XNAColors.colors[color]

    else
        try
            return ((Ahorn.argb32ToRGBATuple(parse(Int, replace(color, "#" => ""), base=16))[1:3] ./ 255)..., 1.0)

        catch

        end
    end

    return (1.0, 1.0, 1.0, 1.0)
end

const colors = sort(collect(keys(Ahorn.XNAColors.colors)))

const placements = Ahorn.PlacementDict(
    "Water (Pandora's Box)" => Ahorn.EntityPlacement(
        ColoredWater,
        "rectangle"
    ),
)

Ahorn.editingOptions(entity::ColoredWater) = Dict{String, Any}(
    "color" => colors
)

Ahorn.minimumSize(entity::ColoredWater) = 8, 8
Ahorn.resizable(entity::ColoredWater) = true, true

Ahorn.selection(entity::ColoredWater) = Ahorn.getEntityRectangle(entity)

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::ColoredWater, room::Maple.Room)
    x = Int(get(entity.data, "x", 0))
    y = Int(get(entity.data, "y", 0))

    width = Int(get(entity.data, "width", 32))
    height = Int(get(entity.data, "height", 32))
    rawColor = get(entity.data, "color", "LightSkyBlue")
    color = getColor(rawColor)

    Ahorn.drawRectangle(ctx, 0, 0, width, height, color .* (1.0, 1.0, 1.0, 0.4), color)
end

end

File: PandorasBox/Loenn/entities/colored_water.lua
local xnaColors = require("consts.xna_colors")
local utils = require("utils")

local water = {}


local function getEntityColor(entity)
    local rawColor = entity.color or "LightSkyBlue"
    local color = utils.getColor(rawColor) or xnaColors.LightSkyBlue

    return color
end

water.name = "pandorasBox/coloredWater"
water.depth = 0
water.fieldOrder = {
    "x", "y", "width", "height",
    "color",
    "hasTop", "hasBottom", "hasLeft", "hasRight",
    "hasTopRays", "hasBottomRays", "hasLeftRays", "hasRightRays",
    "canJumpOnSurface"
}
water.fieldInformation = {
    color = {
        fieldType = "color",
        allowXNAColors = true,
    }
}
water.placements = {
    name = "water",
    data = {
        width = 8,
        height = 8,

        hasTop = true,
        hasBottom = false,
        hasLeft = false,
        hasRight = false,

        hasTopRays = true,
        hasBottomRays = true,
        hasLeftRays = true,
        hasRightRays = true,

        canJumpOnSurface = true,

        color = "LightSkyBlue"
    }
}

water.fillColor = function(room, entity)
    local color = getEntityColor(entity)

    return {color[1] * 0.3, color[2] * 0.3, color[3] * 0.3, 0.6}
end

water.borderColor = function(room, entity)
    local color = getEntityColor(entity)

    return {color[1] * 0.8, color[2] * 0.8, color[3] * 0.8, 0.8}
end

return water